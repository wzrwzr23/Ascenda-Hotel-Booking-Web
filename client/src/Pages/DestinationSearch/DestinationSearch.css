h2 {
    color: white;
    margin: 10px 10px 10px 10px;
}
span {
    color: white;
    margin: 10px 10px 10px 10px;
}

.destformss{
    width: 370px;
    left: 150px;
    top: 480px;
    width: 400px;
    position: absolute;
}

.desti-container {
    display: flex;
    justify-content: center;
    /* margin-top: 20px;
    padding-bottom: 20px; */
    position: absolute;
    left: 150px;
    top: 480px;
    width: 400px;
   
}

/* #usersearch input {
    width: 100%;
} */

.searchBtn{
    display: flex;
    justify-content: center;
    background-color: #bc9a39;  
    color: #e4d2a0;
    text-decoration: none;
    padding:6px;
}

.searchcity{
    width: max-content;
}

.searchItems{
    flex: 1;
    background-color: rgba(112, 73, 119, 0.9);
    padding: 10px;
    border-radius: 10px;
    position: sticky;
    top: 10px;
    height: max-content;
    flex-direction: column;
}

input {
    width: 40%;
    margin: 10px 10px 10px 10px;
    height: 30px;
    padding-left: 10px;
}
.usersearch {
    width: 87%;
}

button {
    width: 100%;
    height: 30px;
    margin-top: 10px;
    cursor: pointer;
    border: none;
    color: white;
    background-color: #bc9a39;
}
button:hover {
    background-color: #e4d2a0;
}

/* .data-space {
    display: inline-block;
    position: relative;
    width: max-content;
} */

.lsItem{
    display: flex;
    flex-direction: column;
    gap: 5px;
    margin-bottom: 10px;
}
  
  
.lsItem>span{
    height: 30px;
    padding: 5px;
    background-color: white;
    display: flex;
    align-items: center;
    cursor: pointer;
}

.headerSearchText {
    margin-left: 10px;
}

/* .listItem:hover {
    background-color: #eccb72;
    transform: scale(1.2);
    left: 7%;
    width: max-content;
} */

/* .date {
    position: absolute;
    top: 100%;
    color: black;
    left: 6%;
} */
  
.options {
    flex: 1;
    flex-direction: column;
    background-color: rgba(254, 187, 2, 0);
    position: relative;
    top: 30%;
    width: fit-content;
    box-shadow: none;
}

.optionItem {
    width: 165px;
    display: flex;
    justify-content: space-between;
    margin: 10px;
}

.optionCounter {
    display: flex;
    align-items: center;
    gap: 10px;
    font-size: 12px;
    color: black;
}

.optionCounterButton {
    width: 20px;
    height: 20px;
    border: 1px solid #0071c2;
    color: rgba(254, 187, 2, 0.9);
    cursor: pointer;
    background-color: #0071c2;
}

.optionCounterButton:disabled {
    cursor: not-allowed;
}