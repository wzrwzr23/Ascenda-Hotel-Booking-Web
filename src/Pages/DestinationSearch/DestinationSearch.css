h2 {
    color: white;
    margin: 10px 10px 10px 10px;
}
span {
    color: white;
    margin: 10px 10px 10px 10px;
}
#destwrap{
    overflow: hidden;
    align-items: center;
}
.dest-container {
    background-color: rgba(254, 187, 2, 0.7);
    width: max-content;
    align-items: center;
    justify-content: center;
    position: absolute;
    height: 380px;
    left: 150px;
    top: 480px;
    overflow: hidden;
}

/* #usersearch input {
    width: 100%;
} */

input {
    width: 40%;
    margin: 12px 12px 12px 12px;
    height: 30px;
    padding-left: 10px;
}
#usersearch{
    width: 85%;
}
a {
    text-decoration: none;
    color: white;
    background-color: #7b5c06;
    width: 100%;
    height: 45px;
    margin-top: 3px;
    cursor: pointer;
    border: none;
    text-align: center;
    display: inline-block;
    padding-top: 10px;
    padding-bottom: 10px;
}
/* .submitDest {
    width: 100%;
    height: 45px;
    margin-top: 32px;
    cursor: pointer;
    border: none;
    color: white;
    background-color: #bc9a39;

} */
a:hover {
    background-color: #003580;
}

