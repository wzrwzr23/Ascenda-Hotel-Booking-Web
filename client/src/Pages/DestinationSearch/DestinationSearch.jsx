import Footer from '../../Components/Footer/Footer'
import Header from '../../Components/Header/Header'
import Navbar from '../../Components/Navbar/Navbar'
import React from 'react'
import { useContext, useState } from 'react'
import destData from '../../destinations.json'
import './DestinationSearch.css'
import { format } from "date-fns";
import { DateRange } from 'react-date-range'
import { SearchContext } from '../../Context/SearchContext'
import { useNavigate } from "react-router-dom";
import { Link } from 'react-router-dom';

// import './DestinationSearch.css'

const DestinationSearch = () => {
  const [userDest, setUserDest] = useState('')
  const [filteredDest, setFilteredDest] = useState([])
  const [checkInDate, setCheckInDate] = useState(null)
  const [checkOutDate, setCheckOutDate] = useState(null)
  const [numChild, setNumChild] = useState()
  const [numAdult, setNumAdult] = useState()
  const [numRoom, setNumRoom] = useState()
  const [destId, setDestId] = useState()
  const [destination, setDestination] = useState("");
  const [openDate, setOpenDate] = useState(false);
  const [dates, setDates] = useState([
    {
      startDate: new Date(),
      endDate: new Date(),
      key: "selection",
    },
  ]);
  const [openOptions, setOpenOptions] = useState(false);
  const [options, setOptions] = useState({
    guest: 1,
    // children: 0,
    room: 1,
  });
  const handleOption = (name, operation) => {
    setOptions((prev) => {
      return {
        ...prev,
        [name]: operation === "i" ? options[name] + 1 : options[name] - 1,
      };
    });
  };

  const {dispatch} = useContext(SearchContext)
  const navigate = useNavigate();

  const handleSearch = async () => {
    // await this.timeout(6500)
    const uid = onSearch()
    setDestId("yo", uid);
    console.log(uid)
    const url = `/hotelsearch/${uid}`;
    // const url = "/login"
    dispatch({type:"NEW_SEARCH", payload:{destination, dates, options}})
    navigate(url, { state: { destination, dates, options } });
  };

  const searchDest = (event) => {
    const userSearch = event
    const lowercase = userSearch.toLowerCase()
    setUserDest(userSearch)
    const filterDest = destData.filter((item) => {
      const destTerm = item.term.toLowerCase()
      return lowercase && destTerm.startsWith(lowercase) && destTerm !== lowercase
    })
    if (userSearch === '') {
      setFilteredDest([])
    } else {setFilteredDest(filterDest.slice(0, 10))}
  }
  // function onSearch(userDest) {
  //   if (userDest === '') {
  //     alert('please fill in all input fields.')
  //   } else {
  //     return getUID(userDest, destData)
  //   }
  // }
  // const onSearch = () => {
  //   if (userDest) {
  //     getUID(userDest, destData)
  //   }
  // }

  function onSearch(){
    var uid;
    var dict = {};
    if (userDest) {
      uid = getUID(userDest, destData)
    }
    
    var checkindate = JSON.stringify(dates[0].startDate).slice(1, 11)
    var checkoutdate = JSON.stringify(dates[0].endDate).slice(1, 11)
    var numguest = options.guest
    var numroom = options.room
    dict['Check-In Date'] = checkindate
    dict['Check-Out Date'] = checkoutdate
    dict['Number of Guests'] = numguest
    dict['Number of Rooms'] = numroom
    alert(JSON.stringify(dict))
  }

  function getUID(value, file) {
    for (var i=0; i<file.length; i++) {
      var obj = file[i]
      if (value === obj.term) {
        setDestId(obj.uid)
        console.log("nami",destId)
        return obj.uid;
      }
    }
    return null;
  }


  return (
    // <div>Destination Search</div>
    <section className='destform'>
      <div className='dest-container'>
        <div className="destWrapper">
        <div className="searchItems">
        <h2>Enjoy Your Stay</h2>  
        <span>Search and Book Hotel</span>
          <div className='usersearch'>
            <input type='text' placeholder='Search City' name='' id='' value={userDest} onChange={(e) => {searchDest(e.target.value); setDestination(e.target.value);}} data-testid='searchinput'/>
          </div>
          <div className='dropdown' data-testid="filter-dest">
            {filteredDest.map((item) => {return (
              <div  className='dropdown-row' style={{color: 'black'}} onClick={() => searchDest(item.term)}>{item.term}</div>
            )})}
          </div>
          <div className='data-space'>
            <span
              onClick={() => setOpenDate(!openDate)}
              className="headerSearchText"
            >{`${format(dates[0].startDate, "MM/dd/yyyy")} to ${format(
                dates[0].endDate,
                "MM/dd/yyyy"
              )}`}</span>
              {openDate && (
                <DateRange
                  editableDateInputs={true}
                  onChange={(item) => setDates([item.selection])}
                  moveRangeOnFirstSelection={false}
                  ranges={dates}
                  className="date"
                  minDate={new Date()}
                />
              )}
          </div>
          <div className="options">
            <div className="optionItem">
              <span className="optionText">Guests</span>
              <div className="optionCounter">
                <button id='guest-count'
                  disabled={options.guest <= 1}
                  className="optionCounterButton"
                  data-testid='guestminus'
                  onClick={() => handleOption("guest", "d")}
                >
                  -
                </button>
                <span className="optionCounterNumber" data-testid='guestno'>
                  {options.guest}
                </span>
                <button
                  className="optionCounterButton"
                  data-testid='guestplus'
                  onClick={() => handleOption("guest", "i")}
                >
                  +
                </button>
              </div>
            </div>
            {/* <div className="optionItem">
              <span className="optionText">Children</span>
              <div className="optionCounter"> */}
                {/* <button
                  disabled={options.children <= 0}
                  className="optionCounterButton"
                  onClick={() => handleOption("children", "d")}
                >
                  -
                </button> */}
                {/* <span className="optionCounterNumber">
                  {options.children}
                </span>
                <button
                  className="optionCounterButton"
                  onClick={() => handleOption("children", "i")}
                >
                  +
                </button>
              </div>
            </div> */}
            <div className="optionItem">
              <span className="optionText">Room</span>
              <div className="optionCounter">
                <button
                  disabled={options.room <= 1}
                  className="optionCounterButton"
                  data-testid='roomminus'
                  onClick={() => handleOption("room", "d")}
                >
                  -
                </button>
                <span className="optionCounterNumber" data-testid='roomno'>
                  {options.room}
                </span>
                <button
                  className="optionCounterButton"
                  data-testid='roomplus'
                  onClick={() => handleOption("room", "i")}
                >
                  +
                </button>
              </div>
            </div>
          </div>
          {/* <Link to={`/hotelsearch/${destId}`}>
              <button className="siCheckButton" onClick={handleSearch}>See availability</button>
          </Link> */}
          {/* <button onClick={handleSearch}>Search</button> */}
        {/*<button type='Submit' className='submitDest' data-testid='submit'*/}
      <a className='searchBtn' href={'/hotelsearch/'+destId} onClick={onSearch} data-testid='submit'>Search</a>
        {/*</button>*/}
        </div>
        </div>
      </div>
    </section>
  ) 
}

export default DestinationSearch