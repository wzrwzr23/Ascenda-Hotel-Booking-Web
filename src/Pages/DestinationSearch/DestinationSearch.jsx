import Footer from '../../Components/Footer/Footer'
import Header from '../../Components/Header/Header'
import Navbar from '../../Components/Navbar/Navbar'
import React from 'react'
import { useState } from 'react'
import destData from '../../destinations.json'
import './DestinationSearch.css'
import { set } from 'date-fns'

// import './DestinationSearch.css'

const DestinationSearch = () => {
  const [userDest, setUserDest] = useState('')
  const [filteredDest, setFilteredDest] = useState([])
  const [checkInDate, setCheckInDate] = useState(null)
  const [checkOutDate, setCheckOutDate] = useState(null)
  const [numChild, setNumChild] = useState()
  const [numAdult, setNumAdult] = useState()
  const [numRoom, setNumRoom] = useState()
  const [destId, setDestId] = useState()

  const searchDest = (event) => {
    const userSearch = event
    const lowercase = userSearch.toLowerCase()
    setUserDest(userSearch)
    const filterDest = destData.filter((item) => {
      const destTerm = item.term.toLowerCase()
      return lowercase && destTerm.startsWith(lowercase) && destTerm !== lowercase
    })
    if (userSearch === '') {
      setFilteredDest([])
    } else {setFilteredDest(filterDest.slice(0, 10))}
  }
  // function onSearch(userDest) {
  //   if (userDest === '') {
  //     alert('please fill in all input fields.')
  //   } else {
  //     return getUID(userDest, destData)
  //   }
  // }
  const onSearch = () => {
    if (userDest) {
      getUID(userDest, destData)
    }
  }

  function getUID(value, file) {
    for (var i=0; i<file.length; i++) {
      var obj = file[i]
      if (value === obj.term) {
        setDestId(obj.uid)
      }
    }
  }


  return (
    // <div>Destination Search</div>
    <section className='destform'>
      <div className='dest-container'>
        <h2>Enjoy Your Stay</h2>  
        <span>Search and Book Hotel</span>

        <form action=''>
          <div className='usersearch'>
            <input type='text' placeholder='Search City' name='' id='' value={userDest} onChange={(e) => searchDest(e.target.value)} data-testid='searchinput'/>
          </div>
          <div className='dropdown' data-testid="filter-dest">
            {filteredDest.map((item) => {return (
              <div  className='dropdown-row' style={{color: 'black'}} onClick={() => searchDest(item.term)}>{item.term}</div>
            )})}
          </div>
          <div className='data-space'>
            <input type='date' placeholder='Check In' value={checkInDate} onChange={(e) => setCheckInDate(e.target.value)}/>
            <input type='date' placeholder='Check Out' value={checkOutDate} onChange={(e) => setCheckOutDate(e.target.value)}/>
          </div>
          <div className='data-space'>
            <input type='number' placeholder='Number of Children' min='0' value={numChild} onChange={(e) => setNumChild(e.target.value)}/>
            <input type='number' placeholder='Number of Adults' min='0' value={numAdult} onChange={(e) => setNumAdult(e.target.value)}/>
          </div>
          <input type='number' placeholder='Number of Rooms' className='room' min='1' value={numRoom} onChange={(e) => setNumRoom(e.target.value)}/>
        </form>
        <a href={'/hotelsearch/'+destId} onClick={onSearch}>Search</a>
      </div>
    </section>
  ) 
}

export default DestinationSearch
