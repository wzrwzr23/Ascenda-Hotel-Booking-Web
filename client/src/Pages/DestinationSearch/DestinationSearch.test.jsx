import React from "react";
import DestinationSearch from "./DestinationSearch";
import { render, fireEvent, queryByTestId, getByTestId, getByPlaceholderText} from "@testing-library/react"
import { BrowserRouter } from "react-router-dom";
import user from '@testing-library/user-event'

it('renders correctly', () => {
    const {getByTestId, queryByPlaceholderText} = render(<BrowserRouter><DestinationSearch/></BrowserRouter>);
    // expect(getByTestId('submit')).toBeTruthy;
    expect(queryByPlaceholderText('Search City')).toBeTruthy;
})

describe('Destination search input: Typing', () => {
    test('search box updates on change', () => {
        const {queryByPlaceholderText} = render(<BrowserRouter><DestinationSearch/></BrowserRouter>);
        const searchInput = queryByPlaceholderText('Search City');
        fireEvent.change(searchInput, {target: {value: "test"}});
        // userEvent.type(searchInput, 'test')
        expect(searchInput.value).toBe("test");
    });
})

describe('Guest count (+)', () => {
    test('guest count updates on change', () => {
        const {getByTestId} = render(<BrowserRouter><DestinationSearch/></BrowserRouter>)
        const guestcount = getByTestId('guestno')
        fireEvent.click(getByTestId('guestplus'))
        expect(guestcount.value).not.toBe(1)
    })
})

describe('Guest count (-)', () => {
    test('guest count does not updates on change', () => {
        const {getByTestId} = render(<BrowserRouter><DestinationSearch/></BrowserRouter>)
        const guestcount = getByTestId('guestno')
        fireEvent.click(getByTestId('guestminus'))
        expect(guestcount.value).toBe(undefined)
    })
})

describe('Room count (+)', () => {
    test('guest count updates on change', () => {
        const {getByTestId} = render(<BrowserRouter><DestinationSearch/></BrowserRouter>)
        const guestcount = getByTestId('roomno')
        fireEvent.click(getByTestId('roomplus'))
        expect(guestcount.value).not.toBe(1)
    })
})

describe('Guest count (-)', () => {
    test('guest count does not updates on change', () => {
        const {getByTestId} = render(<BrowserRouter><DestinationSearch/></BrowserRouter>)
        const guestcount = getByTestId('roomno')
        fireEvent.click(getByTestId('roomminus'))
        expect(guestcount.value).toBe(undefined)
    })
})

describe('Search button', () => {
    test('empty destination does not trigger search(to give uid)', () => {
        const onSearch = jest.fn()
        const {getByTestId, queryByPlaceholderText} = render(<BrowserRouter><DestinationSearch onSearch={onSearch}/></BrowserRouter>)
        fireEvent.click(getByTestId('submit'))
        expect(onSearch).not.toHaveBeenCalled()
    })
})

// DELETE IF DK //
// describe('Search button', () => {
//     test('onSearch is called when all fields are filled', () => {
//         const{queryByPlaceholderText} = render(<BrowserRouter><DestinationSearch/></BrowserRouter>)
//         const userDest = queryByPlaceholderText('Search City')
//         user.type(userDest, 'Indochina War Memorial, Frejus, France')
//         user.click(getByTestId('guest-count'))
//     })
// })


// describe('Destination Result List: Clicking', () => {
//     test('results on the dropdown list can be clicked', () => {
//         const searchDest = jest.fn()
//         const {getByTestId} = render(<BrowserRouter><DestinationSearch searchDest={searchDest}/></BrowserRouter>);
//         fireEvent.click(getByTestId('filter-dest'));
//         expect(searchDest).toHaveBeenCalled()

//     })
// })

// describe('Destination search input: Empty input', () => {
//     test('does not show any destinations', () => {
//         const onSearch = jest.fn()
//         const {getByTestId} = render(<DestinationSearch/>);
//         fireEvent.change(getByTestId('searchinput'), {target: {value: ""}});
//         expect(onSearch).not.toHaveBeenCalled;
//     })
// })

// describe('Destination search input: Destination clicked', () => {
//     test('search box updates', () => {
//         // const onSearch = jest.fn();
//         const {getByTestId} = render(<DestinationSearch/>);
//         const searchInput = getByTestId('searchinput');
//         fireEvent.click(getByTestId('searchinput'));
//         expect(searchInput.value).toBe();
//     })
// })