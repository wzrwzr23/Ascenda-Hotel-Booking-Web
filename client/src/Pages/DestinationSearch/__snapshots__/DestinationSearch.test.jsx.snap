// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<section
  className="destformss"
>
  <div
    className="searchItems"
  >
    <h2>
      Enjoy Your Stay
    </h2>
    <span>
      Search and Book Hotel
    </span>
    <div
      className="usersearch"
    >
      <input
        className="searchcity"
        data-testid="searchinput"
        id="input"
        name=""
        onChange={[Function]}
        placeholder="Search City"
        type="text"
        value=""
      />
    </div>
    <div
      className="dropdown"
      data-testid="filter-dest"
    />
    <div
      className="listItem"
    >
      <span
        className="headerSearchText"
        data-testid="userdate"
        id="date_click"
        onClick={[Function]}
        style={
          Object {
            "color": "white",
            "fontWeight": "bold",
          }
        }
      >
        08/10/2022 to 08/11/2022
      </span>
    </div>
    <div
      className="options"
    >
      <div
        className="optionItem"
      >
        <span
          className="optionText"
        >
          Guests
        </span>
        <div
          className="optionCounter"
        >
          <button
            className="optionCounterButton"
            data-testid="guestminus"
            disabled={true}
            id="guest-count-d"
            onClick={[Function]}
          >
            -
          </button>
          <span
            className="optionCounterNumber"
            data-testid="guestno"
          >
            1
          </span>
          <button
            className="optionCounterButton"
            data-testid="guestplus"
            id="guest-count-i"
            onClick={[Function]}
          >
            +
          </button>
        </div>
      </div>
      <div
        className="optionItem"
      >
        <span
          className="optionText"
        >
          Room
        </span>
        <div
          className="optionCounter"
        >
          <button
            className="optionCounterButton"
            data-testid="roomminus"
            disabled={true}
            id="room-count-d"
            onClick={[Function]}
          >
            -
          </button>
          <span
            className="optionCounterNumber"
            data-testid="roomno"
          >
            1
          </span>
          <button
            className="optionCounterButton"
            data-testid="roomplus"
            id="room-count-i"
            onClick={[Function]}
          >
            +
          </button>
        </div>
      </div>
    </div>
    <button
      data-testid="submit"
      id="search_button"
      onClick={[Function]}
    >
      Search
    </button>
  </div>
</section>
`;
